DOCS-762: Add custom API creation info (#1535)
viamrobotics/docs: docs/extend/modular-resources/create/create-subtype.md

---
title: "Define a New Resource Subtype"
linkTitle: "New API Subtype"
weight: 30
type: "docs"
tags: ["rdk", "extending viam", "modular resources", "API"]
description: "Define an API for a new type of resource."
no_list: true
---

You need to define a new resource *subtype* (and an API for that subtype) if:

- You have a {{% glossary_tooltip term_id="resource" text="resource" %}} that does not fit into any of the existing [component](/components/) or [service](/services/) subtypes.

> - You have a resource that could fit into an existing subtype, but you want to define an API with different endpoints than the ones in the standard Viam [API](/program/apis/) for that subtype.

Reviewer: ```suggestion
- You have a resource that could fit into an existing subtype, but you want to define an API with different methods than the ones in the standard Viam [API](/program/apis/) for that subtype.
```

And should we say "standard VIam API" or "existing APIs" - as soon they could pick from APIs from third parties in the registry?

Author: That's a great point, but what's wrong with endpoints?

Author: ```suggestion
- You have a resource that could fit into an existing subtype, but you want to define an API with different methods and messages than the ones in the existing [APIs](/program/apis/) for that subtype.
```

> For example, you have a [sensor](/components/sensor/) and you want to define a `Calibrate` endpoint.

Reviewer: ```suggestion
For example, you have a [sensor](/components/sensor/) and you want to define a `Calibrate` method.
```

{{% alert title="Tip" color="tip" %}}

If you want to use most of an existing API but need just a few other functions, try using the [`DoCommand`](/program/apis/#docommand) endpoint and [extra parameters](/program/use-extra-params/) to add custom functionality to an existing subtype.

{{% /alert %}}

> ## How to define your new resource API

Reviewer 2: ```suggestion
## Define your new resource API
```

      - [Example service in Python](https://github.com/viam-labs/speech/blob/main/src/speech/api.py)

5. In the root directory of your module, you need to generate some boilerplate files.
  You will typically need the following three files for most modules, though different files are required for some advanced use cases.
  See the [Buf documentation](https://buf.build/docs/generate/usage/) for instructions.

    - [<file>buf.yaml</file>](https://buf.build/docs/configuration/v1/buf-gen-yaml/)
    - [<file>buf.gen.yaml</file>](https://buf.build/docs/configuration/v1/buf-gen-yaml/)
    - [<file>buf.lock</file>](https://buf.build/docs/configuration/v1/buf-lock/)

6. In the <file>/src/</file> directory of your module, use the protobuf compiler to [generate](https://buf.build/docs/tutorials/getting-started-with-buf-cli/#generate-code) all other necessary protocol buffer code, based on the `<subtype name>.proto` file you wrote.

    - [Example generated files for a Python-based service](https://github.com/viam-labs/speech/tree/main/src/proto)
      - (<file>speech.proto</file> was manually written, not generated)

7. Your resource API is defined.

>   Now you can [create a new model](/extend/modular-resources/create/#code-a-new-resource-model) that implements your new API.

Reviewer 2: Can we add some next steps here? That's the only thing missing I think
